packages/react-dom/src/__tests__/ReactDOMNativeEventHeuristic-test.js

```diff
@@ -0,0 +1,230 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @emails react-core
+ */
+
+'use strict';
+
+let React;
+
+let ReactDOM;
+let Scheduler;
+
+describe('ReactDOMNativeEventHeuristic-test', () => {
+  let container;
+
+  beforeEach(() => {
+    jest.resetModules();
+    container = document.createElement('div');
+    React = require('react');
+    ReactDOM = require('react-dom');
+    Scheduler = require('scheduler');
+
+    document.body.appendChild(container);
+  });
+
+  afterEach(() => {
+    document.body.removeChild(container);
+  });
+
+  function dispatchAndSetCurrentEvent(el, event) {
+    try {
+      window.event = event;
+      el.dispatchEvent(event);
+    } finally {
+      window.event = undefined;
+    }
+  }
+
+  // @gate experimental
+  // @gate enableDiscreteEventMicroTasks && enableNativeEventPriorityInference
+  it('ignores discrete events on a pending removed element', async () => {
+    const disableButtonRef = React.createRef();
+    const submitButtonRef = React.createRef();
+
+    function Form() {
+      const [active, setActive] = React.useState(true);
+
+      React.useLayoutEffect(() => {
+        disableButtonRef.current.onclick = disableForm;
+      });
+
+      function disableForm() {
+        setActive(false);
+      }
+
+      return (
+        <div>
+          <button ref={disableButtonRef}>Disable</button>
+          {active ? <button ref={submitButtonRef}>Submit</button> : null}
+        </div>
+      );
+    }
+
+    const root = ReactDOM.unstable_createRoot(container);
+    root.render(<Form />);
+    // Flush
+    Scheduler.unstable_flushAll();
+
+    const disableButton = disableButtonRef.current;
+    expect(disableButton.tagName).toBe('BUTTON');
+
+    // Dispatch a click event on the Disable-button.
+    const firstEvent = document.createEvent('Event');
+    firstEvent.initEvent('click', true, true);
+    expect(() =>
+      dispatchAndSetCurrentEvent(disableButton, firstEvent),
+    ).toErrorDev(['An update to Form inside a test was not wrapped in act']);
+
+    // There should now be a pending update to disable the form.
+    // This should not have flushed yet since it's in concurrent mode.
+    const submitButton = submitButtonRef.current;
+    expect(submitButton.tagName).toBe('BUTTON');
+
+    // Discrete events should be flushed in a microtask.
+    // Verify that the second button was removed.
+    await null;
+    expect(submitButtonRef.current).toBe(null);
+    // We'll assume that the browser won't let the user click it.
```

Comment:
The original test was more interesting here because it fired an event and verified that React ignores it. However, if we fire an event manually on a DOM node, and listen directly on that node with a native listener, it would still fire. So I don't know what we should be testing here. I think just testing that the DOM node was removed is enough because the rest is up to the browser.